File: matiasmasso/Mat.NET/DT2/DTOIncentiu.vb
Public Class DTOIncentiu
    Inherits DTOBaseGuid

    Property FchFrom As Date
    Property FchTo As Date
    Property Product As DTOProduct 'marca per discriminar a qui se li comunica
    Property Channels As List(Of DTODistributionChannel)
    Property Products As List(Of DTOProduct) 'productes als que se li aplicará automaticament la promo
    Property QtyDtos As List(Of DTOQtyDto)

    Property Title As DTOLangText
    Property Excerpt As DTOLangText
    Property Bases As DTOLangText
    Property ManufacturerContribution As String

    Property CliVisible As Boolean
    Property RepVisible As Boolean

    Property OnlyInStk As Boolean
    Property Cod As Cods
    Property Evento As DTOEvento
    Property PdcsCount As Integer
    Property PointsOfSale As List(Of DTOContact)

    <JsonIgnore> Property Thumbnail As Image
    Property Unitats As Integer
    Property Dto As Integer

    Public Enum wellknowns
        NotSet
        FeriasLocales
    End Enum

    Public Enum Cods
        NotSet
        Dto
        FreeUnits
    End Enum

    Public Sub New()
        MyBase.New()
        _Title = New DTOLangText()
        _Excerpt = New DTOLangText
        _Bases = New DTOLangText
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
        _Title = New DTOLangText()
        _Excerpt = New DTOLangText
        _Bases = New DTOLangText
    End Sub

    Shared Function wellknown(id As DTOIncentiu.wellknowns) As DTOIncentiu
        Dim retval As DTOIncentiu = Nothing
        Select Case id
            Case DTOIncentiu.wellknowns.FeriasLocales
                retval = New DTOIncentiu(New Guid("DA758FDB-983F-4A43-B0A9-70DF027CC878"))
        End Select
        Return retval
    End Function


    Shared Function GetDto(oIncentiu As DTOIncentiu, ByVal iQty As Integer) As Decimal
        Dim retval As Decimal
        For Each oItm As DTOQtyDto In oIncentiu.QtyDtos
            If iQty < oItm.Qty Then Exit For
            retval = oItm.Dto
        Next
        Return retval
    End Function


End Class

Public Class DTOIncentiuResult
    Property Incentiu As DTOIncentiu
    Property QtyDto As DTOQtyDto 'acumula les unitats dels albarans i fixa el descompte que li toca
End Class

File: matiasmasso/Mat.NET/DTO/DTOIncentiu.cs
using Newtonsoft.Json;

using System;
using System.Collections.Generic;

namespace DTO
{
    public class DTOIncentiu : DTOBaseGuid
    {
        public DateTime FchFrom { get; set; }
        public DateTime FchTo { get; set; }
        public DTOProduct Product { get; set; } // marca per discriminar a qui se li comunica
        public List<DTODistributionChannel> Channels { get; set; }
        public List<DTOProduct> Products { get; set; } // productes als que se li aplicará automaticament la promo
        public List<DTOQtyDto> QtyDtos { get; set; }

        public DTOLangText Title { get; set; }
        public DTOLangText Excerpt { get; set; }
        public DTOLangText Bases { get; set; }
        public string ManufacturerContribution { get; set; }

        public bool CliVisible { get; set; }
        public bool RepVisible { get; set; }

        public bool OnlyInStk { get; set; }
        public Cods Cod { get; set; }
        public DTOEvento Evento { get; set; }
        public int PdcsCount { get; set; }
        public List<DTOContact> PointsOfSale { get; set; }

        [JsonIgnore]
        public Byte[] Thumbnail { get; set; }
        public int Unitats { get; set; }
        public int Dto { get; set; }

        public enum Wellknowns
        {
            notSet,
            feriasLocales
        }

        public enum Cods
        {
            notSet,
            dto,
            freeUnits
        }

        public DTOIncentiu() : base()
        {
            this.Title = new DTOLangText(base.Guid, DTOLangText.Srcs.IncentiuTitle);
            this.Excerpt = new DTOLangText(base.Guid, DTOLangText.Srcs.IncentiuExcerpt);
            this.Bases = new DTOLangText(base.Guid, DTOLangText.Srcs.IncentiuBases);
        }

        public DTOIncentiu(Guid oGuid) : base(oGuid)
        {
            this.Title = new DTOLangText(base.Guid, DTOLangText.Srcs.IncentiuTitle);
            this.Excerpt = new DTOLangText(base.Guid, DTOLangText.Srcs.IncentiuExcerpt);
            this.Bases = new DTOLangText(base.Guid, DTOLangText.Srcs.IncentiuBases);
        }

        public static DTOIncentiu Wellknown(DTOIncentiu.Wellknowns id)
        {
            DTOIncentiu retval = null;
            switch (id)
            {
                case DTOIncentiu.Wellknowns.feriasLocales:
                    {
                        retval = new DTOIncentiu(new Guid("DA758FDB-983F-4A43-B0A9-70DF027CC878"));
                        break;
                    }
            }
            return retval;
        }


        public static decimal GetDto(DTOIncentiu oIncentiu, int iQty)
        {
            decimal retval = 0;
            foreach (DTOQtyDto oItm in oIncentiu.QtyDtos)
            {
                if (iQty < oItm.Qty)
                    break;
                retval = oItm.Dto;
            }
            return retval;
        }
    }

    public class DTOIncentiuResult
    {
        public DTOIncentiu Incentiu { get; set; }
        public DTOQtyDto QtyDto { get; set; } // acumula les unitats dels albarans i fixa el descompte que li toca
    }
}
